opm/autodiff/SimulatorFullyImplicitBlackoilOutput.hpp

```diff
@@ -375,5 +379,291 @@ namespace Opm
+                const Opm::PhaseUsage& phaseUsage,
+                const Model& model,
+                const RestartConfig& restartConfig,
+                const int reportStepNum) {
+
+
+            std::vector<data::CellData> simProps;
+
+            std::map<const char*, int> outKeywords {
+                {"ALLPROPS", 0},
+
+                // Formation volume factors
+                {"BG", 0},
+                {"BO", 0},
+                {"BW", 0},
+
+                {"CONV", 0}, // < Cells with convergence problems
+                {"DEN", 0},  // < Densities
+
+                // Relative permeabilities
+                {"KRG", 0},
+                {"KRO", 0},
+                {"KRW", 0},
+
+                {"RVSAT", 0}, // < Vaporized gas/oil ratio
+                {"RSSAT", 0}, // < Dissolved gas/oil ratio
+
+                {"NORST", 0}, // < Visualization restart file only
+                {"PBPD", 0},  // < Bubble point and dew point pressures
+                {"VISC", 0}   // < Viscosities
+            };
+
+            //Get the value of each of the keys
+            for (auto& keyValue : outKeywords) {
+                keyValue.second = restartConfig.getKeyword(keyValue.first, reportStepNum);
+            }
+
+            //Postprocess some of the special keys
+            if (outKeywords["ALLPROPS"] > 0) {
+                //ALLPROPS implies KRO,KRW,KRG,xxx_DEN,xxx_VISC,BG,BO (xxx= OIL,GAS,WAT)
+                outKeywords["BG"] = std::max(outKeywords["BG"], 1);
+                outKeywords["BO"] = std::max(outKeywords["BO"], 1);
+                outKeywords["BW"] = std::max(outKeywords["BW"], 1);
+
+                outKeywords["DEN"] = std::max(outKeywords["DEN"], 1);
+
+                outKeywords["KRG"] = std::max(outKeywords["KRG"], 1);
+                outKeywords["KRO"] = std::max(outKeywords["KRO"], 1);
+                outKeywords["KRW"] = std::max(outKeywords["KRW"], 1);
+
+                outKeywords["VISC"] = std::max(outKeywords["VISC"], 1);
+            }
+
+            const std::vector<typename Model::ReservoirResidualQuant>& rq = model.getReservoirResidualQuantities();
+
+
+            //Get shorthands for water, oil, gas
+            const int aqua_active = phaseUsage.phase_used[Opm::PhaseUsage::Aqua];
+            const int liquid_active = phaseUsage.phase_used[Opm::PhaseUsage::Liquid];
+            const int vapour_active = phaseUsage.phase_used[Opm::PhaseUsage::Vapour];
+
+            const int aqua_idx = phaseUsage.phase_pos[Opm::PhaseUsage::Aqua];
+            const int liquid_idx = phaseUsage.phase_pos[Opm::PhaseUsage::Liquid];
+            const int vapour_idx = phaseUsage.phase_pos[Opm::PhaseUsage::Vapour];
+
+
+            /**
+             * Formation volume factors for water, oil, gas
+             */
+            if (aqua_active && outKeywords["BW"] > 0) {
+                simProps.emplace_back(
+                        "1OVERBW",
+                        Opm::UnitSystem::measure::volume,
```

Comment:
I am not sure volume is correct here, it's a ratio of volumes: surface volume / reservoir volume.